### ISSUE TITLE
[new rule] Disallow overwriting methods with jest.fn()

### ISSUE THREAD

**Issue by fadookie** [jest-community/eslint-plugin-jest#185]
I have seen a lot of my colleagues write code like this (based on many examples I've seen in the jest community):
```
Date.now = jest.fn().mockReturnValue(new Date('2016'));
```

The problem with overwriting methods on singleton/global objects like this is that jest has no way to restore the mock after the test completes. This can lead to test interactions and nasty heisenbugs due to how jest runs tests in parallel.

A better way is via `spyOn`, ex:
```
jest.spyOn(Date, 'now').mockReturnValue(new Date('2016'));
```

This allows jest to track mocks and un-mock them later via the `restoreMocks` config flag or an explicit call to `jest.restoreAllMocks()`.

Given that I don't think the linter can reasonably determine whether or not a given object is a singleton, how about a rule that disallows any overwriting of object properties with `jest.fn()`?
So something like this should still be valid:
```
const mockObject = { doSomething: jest.fn() };
```
but not the first example.

I've never written an eslint rule but if I can find time I might take a crack at this. Anyone else is welcome to also.

**Comment by SimenB** [comment]
Great idea!

**Comment by fadookie** [comment]
@hanneslund @SimenB thanks so much! We will look into using this new rule soon.

**Comment by garyking** [comment]
If we're using `jest.fn()` in a mock file (in a `__mocks__` dir), then how do we replace that with `jest.spyOn`? Especially since my understanding is that `spyOn` is only restored if it's declared within a test?

**Comment by SimenB** [comment]
Why would you want to restore something if it's in a mock file? Or is because it's more typing that you don't _need_ restoration?

**Comment by garyking** [comment]
No, I'm saying that this rule warns to replace `jest.fn` with `jest.spyOn` in a mock file, when it shouldn't.

**Comment by kleptog** [comment]
We're having trouble with this rule with the following code which uses $notify from the vue-notification plugin.
```
wrapper.vm.$notify = jest.fn(() => Promise.resolve());
```
This gets converted to:
```
jest.spyOn(wrapper.vm, '$notify').mockImplementation(() => Promise.resolve());
```
However this fails with:
```
    Cannot spy the $notify property because it is not a function; undefined given instead
```
Because the plugin doesn't exist in test cases. For now we simply disable this rule here, but what is the correct way of handling this? Seems to me that spyOn could simply be made to work with undefined values?

**Comment by SimenB** [comment]
Actually had the same issue at work with a `jest.spyOn(window, 'fetch')` vs `window.fetch = jest.fn()`.

I'm not sure if it's correct of Jest to add a function if `objectPassed[propertyPassed]` is undefined. But I'm also not sure what the best way forward here is...

@thymikee @rickhanlonii thoughts?

**Comment by thymikee** [comment]
I think we should keep the `spyOn` behavior as it is, because it's correct. Can ESLint check if `objectPassed[propertyPassed]` exists, and if it isn't allow `jest.fn`?

**Comment by SimenB** [comment]
>Can ESLint check if `objectPassed[propertyPassed]` exists, and if it isn't allow `jest.fn`

No, we cannot know that when doing static analysis

**Comment by thymikee** [comment]
Such a pain. I wonder if we could use something like https://github.com/kentcdodds/babel-plugin-preval, quite a bit of overhead tho

**Comment by kleptog** [comment]
What is needed is a variant of spyOn() that doesn't call the original function, which means it doesn't require the original function to exist in the first place. Something like:
```
spy = jest.replaceWithMock(windows, 'fetch', () => "foo")
expect(spy.mock.calls.length).toBe(1)
```

**Comment by rickhanlonii** [comment]
Fwiw by default spyOn doesn't call the original function

**Comment by rickhanlonii** [comment]
Sorry, I'm wrong, I was thinking of Jasmine spyOn 😅

**Comment by fadookie** [comment]
It's an interesting question. If there are only a handful of instances of this I suppose you could just add one-off exceptions via `// eslint-disable-line jest/prefer-spy-on` or somesuch. It doesn't seem that static analysis of a dynamic language like JS is really going to be able to distinguish these cases with such nuance.

I wonder if the linter would catch this form:
```javascript
Object.assign(wrapper.vm, { '$notify': jest.fn(() => Promise.resolve()) });
```
Kind of pointless perhaps, but maybe more obvious that you are jumping through a hoop for some reason to avoid direct property assignment?
